### PR TITLE
test: skip instead of panicking without LLM credentials; document two dead knobs

### PR DESCRIPTION
**Pull request by GraDKh**
Closes [SDK-627](https://linear.app/cognee/issue/SDK-627). Item 6 of #212 (`B6`, `C4`, `B5`). No runtime behaviour changes.

## 1. Fifteen integration tests hard-panicked without credentials

`create_llm_from_env` replayed a cassette under `COGNEE_TEST_REPLAY` and otherwise called straight into `require_env`, which panics. **14 call sites / 15 test functions / 8 files / 3 crates** had no guard, so `cargo test -p cognee-cognify` was red for anyone without an `OPENAI_URL`. CI never saw it — it sets `COGNEE_TEST_REPLAY=1`, which takes the cassette path and never reaches the panic.

### Reproducing it needs a step the ticket doesn't mention

The repo has a root `.env` carrying credentials, and `dotenv` finds it from any crate directory — so the bug is **invisible locally** unless that is neutralised. `dotenv` does not override already-set vars, so this reproduces it without touching the file:

```bash
env OPENAI_URL= OPENAI_TOKEN= LLM_ENDPOINT= LLM_API_KEY= COGNEE_TEST_REPLAY= \
  cargo test -p cognee-cognify --test integration_fact_extraction
```

Before: `FAILED. 0 passed; 3 failed` — `❌ Required environment variable 'OPENAI_URL' is not set`.

### Fixed at the helper, and structurally

`create_llm_from_env` now returns `Option<Arc<dyn Llm>>`. The `Option` is the point: a guard that lives in a *separate* call is a guard fifteen tests can forget — and did. Now the type system will not let a caller skip it.

The gate itself is single-sourced as `cognee_test_utils::create_openai_adapter_if_available`. Cassette handling stays in the per-crate helpers deliberately: `cognee-llm/mock` is feature-gated and not enabled for `cognee-test-utils`, so hoisting it would mean a dependency change on a crate every test crate shares — more blast radius than this fix warrants.

Two things fell out:

- `delete`'s inlined copy also `require_env`-ed `OPENAI_MODEL`, so it panicked on a machine with an endpoint and key but **no model**. It now uses the shared helper, which defaults the model. That made a local `require_env` and an import dead — removed.
- In `delete`, `None` propagates out of `setup_infrastructure` with `?`; both callers already treat that as "skip", so no separate guard was needed there.

### Measured

| | before | after |
|---|---|---|
| `integration_fact_extraction`, keyless | 0 passed, **3 failed** | passes (skips) |
| cognify + search + delete, keyless | — | **894 passed, 0 failed** |
| cassette replay (`COGNEE_TEST_REPLAY=1`) | passes | still passes |

The replay check matters: it is how CI runs, so breaking it would have been invisible in the keyless run above.

## 2. Cassette/schema coupling is now documented

`cassette::input_hash` hashes the canonicalized schema. `canonicalize` sorts object keys — so property *order* is normalised away — but it serialises every value, `description` strings included. So **editing a `///` doc comment on any type reachable from the extraction schema invalidates every committed cassette**, and the replay lane fails on a miss.

Recorded next to the existing `///`-vs-`//` note on `KnowledgeGraph::edges`, which is exactly where someone about to make that edit is already reading.

## 3. `chunk_overlap` documented as inert

Verified: **zero references in `crates/chunking/`** — setting it changes no chunk boundary. Python is the same shape: it declares `ChunkConfig.chunk_overlap = 10` and `DefaultChunkEngine` genuinely consumes it, but `get_chunking_engine()` has no callers and the default path is `TextChunker`, which never reads it. (`text_chunker_with_overlap.py` uses a *ratio* — a different, opt-in knob.)

Kept rather than removed, per the ticket's decision: removal is a SemVer break across four public binding surfaces for a knob that costs nothing to keep, and would diverge from Python, which still declares it. The reasoning lives in `CognifyConfig::chunk_overlap`, with a one-line pointer from `python/cognee_py/types.py`, `ts/src/types.ts`, `java/…/CognifyOptions.java` and `capi/include/cognee_sdk.h`.

**Corrected while there:** the existing comment claimed *"Used when chunk_strategy is RECURSIVE or LANGCHAIN"*. That is true of no code path.

## Verification

`scripts/check_all.sh`: all checks passed (Java binding skipped locally — no JDK; CI covers it). Rebased onto `5f0763f8` and re-verified: clippy clean, 894 keyless tests still pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01C1A31FeijBqvdh7YWjNh13

### CODE REVIEW

File: capi/include/cognee_sdk.h [hunk 1]
```diff
@@ -407,7 +407,9 @@ void cg_sdk_add(const CgSdk*        sdk,
  * opts_json may be NULL or a JSON object with optional keys:
  *   "tenant"         — UUID string
  *   "chunkSize"      — integer
- *   "chunkOverlap"   — integer
+ *   "chunkOverlap"   — integer (INERT: nothing in the chunking pipeline
+ *                      reads it, in Rust or Python. Accepted and ignored;
+ *                      kept for API compatibility.)
```
[the remark this answers is not in the document]
Comment (GraDKh) [reply capi/include/cognee_sdk.h]: Fixed in 502de0a7. Corrected on all four binding surfaces: the note now says inert **on the default chunking path** — no chunker reads it, so chunk boundaries are unaffected — and names validation as the one thing that does read it (rejecting overlap >= chunk size).

File: crates/cognify/src/config.rs [hunk 2]
```diff
@@ -104,9 +104,32 @@ pub struct CognifyConfig {
...
-    /// Python default: 10 (from ChunkConfig.chunk_overlap)
-    /// Used when chunk_strategy is RECURSIVE or LANGCHAIN
+    /// Overlap between chunks (in tokens). Python default: 10 (from
+    /// `ChunkConfig.chunk_overlap`).
+    ///
+    /// # Inert — kept deliberately, on both sides
+    ///
+    /// **Nothing reads this.** It is defaulted and validated here, plumbed
+    /// through `Settings`, the CLI and `bindings-common`, and then never
+    /// consulted: `crates/chunking/` contains zero references to it. Setting it
```
[the remark this answers is not in the document]
Comment (GraDKh) [reply crates/cognify/src/config.rs]: Fixed in 502de0a7. You're right that "Nothing reads this" was overstated: `CognifyConfig::validate` rejects `chunk_overlap >= max_chunk_size` at config.rs:698, so an out-of-range value is still an error even though an in-range one has no effect. The doc now scopes the claim to chunking code and calls out validation explicitly, since that is exactly what would surprise someone who read the field as entirely unobservable.

File: crates/cognify/src/fact_extraction/models.rs [hunk 3]
```diff
@@ -128,6 +128,15 @@ pub struct KnowledgeGraph {
     // tolerance it should never aim at. `the_model_facing_schema_carries_no_
     // implementation_detail` below pins that.
     //
+    // It would also INVALIDATE EVERY COMMITTED CASSETTE. `cassette::input_hash`
+    // hashes the canonicalized schema alongside the messages, and while
+    // `canonicalize` sorts object keys — so property *order* does not matter —
+    // it serialises every value, `description` strings included. Editing or
+    // adding a `///` on any type reachable from this schema therefore changes
+    // the hash of every recorded call, and the replay lane
```
[the remark this answers is not in the document]
Comment (GraDKh) [reply crates/cognify/src/fact_extraction/models.rs]: Fixed in 502de0a7. Scoped to structured-output cassettes — `input_hash` appends the canonicalized schema only when one is passed, so plain chat entries hash messages alone and a doc-comment edit cannot disturb them.

File: crates/test-utils/src/lib.rs [hunk 4]
```diff
@@ -209,6 +209,44 @@ pub fn create_openai_adapter_from_env() -> Arc<OpenAIAdapter> {
...
+/// each cassette, runs the tests, then `git add`s the fixture glob and commits
+/// whatever the staged diff holds. A skip there would exit 0 with the files
+/// still deleted, so the workflow would commit the deletions as a "successful"
+/// recording and every later replay run would fail on a cassette miss. Its
+/// credentials come from `OPENAI_TOKEN: ${{ secrets.BASETEN_KEY }}`, which is
+/// empty whenever that secret is absent — so this is reachable, not theoretical.
+///
+/// Before the keyless skip existed, `require_env`'s panic happened to protect
+/// that workflow by failing the step. Panicking here deliberately preserves it.
+pub fn create_openai_adapter_if_available() -> Option<Arc<OpenAIAdapter>> {
```
Comment (GraDKh) [inline crates/test-utils/src/lib.rs:236]: Addressing the suppressed comment on this helper (no thread to reply to).

**Confirmed, and this was a regression this PR introduced — the best catch of the review.** Verified against `.github/workflows/record-cassettes.yml`: it runs `rm -f` on each cassette, then `cargo test`, then `git add '**/tests/fixtures/cassettes/*.json'` and commits whatever the staged diff holds. Its credentials are `OPENAI_TOKEN: ${{ secrets.BASETEN_KEY }}`, so they are empty whenever that secret is absent — reachable, not theoretical.

Before this PR, the missing-credential panic in `require_env` made `cargo test` exit non-zero and abort the step. My keyless skip removed that, so the run would have exited 0 with the files still deleted and committed the deletions as a successful recording, leaving every later replay run to fail on a cassette miss. The old panic was accidentally load-bearing.

Fixed in 502de0a7: this now panics when `COGNEE_RECORD_LLM` is set but no credentials are available, and returns `None` only for the ordinary keyless case. Measured both directions — keyless still skips (3 passed); record-without-credentials fails 3/3, so the step aborts before the commit.

File: crates/cognify/tests/test_utils.rs [hunk 5]
```diff
@@ -38,22 +38,26 @@ pub fn cassette_path(name: &str) -> String {
 ///
 /// Pair this with `MOCK_EMBEDDING=deterministic` for any test that also touches
 /// the vector store, so embeddings are identical across record and replay.
+/// Returns `None` when neither replay nor live credentials are available, so the
+/// caller skips instead of panicking. See
+/// `cognee_test_utils::create_openai_adapter_if_available` for why this is an
+/// `Option` rather than a separate guard the caller has to remember.
 #[allow(dead_code)]
-pub fn create_llm_from_env(cassette_name: &str) -> Arc<dyn Llm> {
+pub fn create_llm_from_env(cassette_name: &str) -> Option<Arc<dyn Llm>> {
```
Comment (GraDKh) [inline crates/cognify/tests/test_utils.rs:46]: Addressing the suppressed comment on this signature (no thread to reply to).

Correct, and fixed in 502de0a7. `docs/build/ci-test-parallelism.md` described Approach E's third branch as "neither → the real adapter", which this PR's change to `Option<Arc<dyn Llm>>` made wrong. It now reads "the real adapter when credentials are available, otherwise `None`, and the caller skips", with a note on why the return type is an `Option` at all — a separate guard is one fifteen tests forgot.

The record branch in that document gained the matching warning too: credentials are required there, and absence panics rather than skipping, for the `record-cassettes.yml` reason in the other suppressed comment.
